### PR TITLE
Upgrade c-ares to 1.13.0

### PR DESCRIPTION
**Pull request by jtattermusch**
-- starting from this version, c-ares has its own support for cmake, switch to using c-ares cmake
-- update ares_build.h and ares_config.h  that's used for bazel build &  Makefile build

Using c-ares' cmake is useful because it supports installing c-ares, which is in turn needed for properly supporting installing gRPC as a cmake package (I have some in-progress work to improve gRPC's cmake build).

### CODE REVIEW

**Comment by jtattermusch** [comment]
@y-zeng  Where can I find a freebsd machine to regenerate the ares_config.h?

**Comment by y-zeng** [comment]
@jtattermusch Thanks for doing this upgrade!
@mehrdada Do you know where we can find a freebsd machine to regenerate the ares_config.h?

File: third_party/cares/config_linux/ares_config.h [hunk 1]
```diff
@@ -46,38 +45,29 @@
...
-#define HAVE_BOOL_T 1
+#define HAVE_BOOL_T
 
-/* Define HAVE_CLOCK_GETTIME_MONOTONIC to 1 if you have the clock_gettime
- * function and monotonic timer.
- *
- * Note: setting HAVE_CLOCK_GETTIME_MONOTONIC causes use of the clock_gettime
- * function from glibc, don't set it to support glibc < 2.17 */
-#ifndef GPR_BACKWARDS_COMPATIBILITY_MODE
-  #define HAVE_CLOCK_GETTIME_MONOTONIC 1
```
Comment (apolcyn) [inline third_party/cares/config_linux/ares_config.h]: I think we need to preserve this manual edit. `HAVE_CLOCK_GETTIME_MONOTONIC` being defined forces a dependency on glibc 1.17+ - it should break a couple of ruby distrib tests.

File: third_party/cares/config_linux/ares_config.h [hunk 2]
```diff
@@ -453,106 +409,15 @@
...
-/* #  undef _ALL_SOURCE */
-#endif
-
-/* Enable large inode numbers on Mac OS X 10.5.  */
-#ifndef _DARWIN_USE_64_BIT_INODE
-# define _DARWIN_USE_64_BIT_INODE 1
-#endif
-
-#ifdef GPR_BACKWARDS_COMPATIBILITY_MODE
-  /* Redefine the fd_set macros for GLIBC < 2.15 support.
```
Comment (apolcyn) [inline third_party/cares/config_linux/ares_config.h]: Same here, I think we also need to preserve this manual edit. Building against latest `FD_SET` macros forces the binary dependency on glibc1.5+ (a couple of ruby distrib tests should also fail on this, https://github.com/grpc/grpc/issues/11346 for context)

File: third_party/cares/config_linux/ares_config.h (hunk 1 above)
Comment (jtattermusch) [reply third_party/cares/config_linux/ares_config.h]: Done.

File: third_party/cares/config_linux/ares_config.h (hunk 2 above)
Comment (jtattermusch) [reply third_party/cares/config_linux/ares_config.h]: Done.

**Comment by jtattermusch** [comment]
@apolcyn, how was the ares_config.h for macos generated? I generated one one my gMac, but I'm getting a build error:
```
mkdir -p `dirname /jenkins/workspace/gRPC_pull_requests_macos/workspace_c_macos_dbg_native/objs/dbg/third_party/cares/cares/ares__timeval.o`
cc -Ithird_party/protobuf/src -Ithird_party/googletest/googletest/include -Ithird_party/googletest/googlemock/include -Ithird_party/boringssl/include -Ithird_party/cares -Ithird_party/cares/cares -g -Wall -Wextra -Werror -Wno-long-long -Wno-unused-parameter -DOSATOMIC_USE_INLINED=1 -O0 -fPIC -I. -Iinclude -I/jenkins/workspace/gRPC_pull_requests_macos/workspace_c_macos_dbg_native/gens -I/opt/local/include -I/usr/local/include -D_DEBUG -DDEBUG -DINSTALL_PREFIX=\"/usr/local\" -DGRPC_TEST_SLOWDOWN_MACHINE_FACTOR=1.000000    -Ithird_party/cares -Ithird_party/cares/cares  -Ithird_party/cares/config_darwin -fvisibility=hidden -D_GNU_SOURCE -DWIN32_LEAN_AND_MEAN -D_HAS_EXCEPTIONS=0 -DNOMINMAX -DHAVE_CONFIG_H -std=c99 -Wsign-conversion -Wconversion -Wshadow -Wextra-semi   -Wno-sign-conversion -Wno-shorten-64-to-32 -Wno-invalid-source-encoding -MMD -MF /jenkins/workspace/gRPC_pull_requests_macos/workspace_c_macos_dbg_native/objs/dbg/third_party/cares/cares/ares__read_line.dep -c -o /jenkins/workspace/gRPC_pull_requests_macos/workspace_c_macos_dbg_native/objs/dbg/third_party/cares/cares/ares__read_line.o third_party/cares/cares/ares__read_line.c
cc -Ithird_party/protobuf/src -Ithird_party/googletest/googletest/include -Ithird_party/googletest/googlemock/include -Ithird_party/boringssl/include -Ithird_party/cares -Ithird_party/cares/cares -g -Wall -Wextra -Werror -Wno-long-long -Wno-unused-parameter -DOSATOMIC_USE_INLINED=1 -O0 -fPIC -I. -Iinclude -I/jenkins/workspace/gRPC_pull_requests_macos/workspace_c_macos_dbg_native/gens -I/opt/local/include -I/usr/local/include -D_DEBUG -DDEBUG -DINSTALL_PREFIX=\"/usr/local\" -DGRPC_TEST_SLOWDOWN_MACHINE_FACTOR=1.000000    -Ithird_party/cares -Ithird_party/cares/cares  -Ithird_party/cares/config_darwin -fvisibility=hidden -D_GNU_SOURCE -DWIN32_LEAN_AND_MEAN -D_HAS_EXCEPTIONS=0 -DNOMINMAX -DHAVE_CONFIG_H -std=c99 -Wsign-conversion -Wconversion -Wshadow -Wextra-semi   -Wno-sign-conversion -Wno-shorten-64-to-32 -Wno-invalid-source-encoding -MMD -MF /jenkins/workspace/gRPC_pull_requests_macos/workspace_c_macos_dbg_native/objs/dbg/third_party/cares/cares/ares__timeval.dep -c -o /jenkins/workspace/gRPC_pull_requests_macos/workspace_c_macos_dbg_native/objs/dbg/third_party/cares/cares/ares__timeval.o third_party/cares/cares/ares__timeval.c
third_party/cares/cares/ares__timeval.c:48:11: error: implicit declaration of function 'clock_gettime' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
  if(0 == clock_gettime(CLOCK_MONOTONIC, &tsnow)) {
          ^
[C]       Compiling third_party/cares/cares/ares_cancel.c
third_party/cares/cares/ares__timeval.c:48:25: error: use of undeclared identifier 'CLOCK_MONOTONIC'
  if(0 == clock_gettime(CLOCK_MONOTONIC, &tsnow)) {
                        ^
2 errors generated.
mkdir -p `dirname /jenkins/workspace/gRPC_pull_requests_macos/workspace_c_macos_dbg_native/objs/dbg/third_party/cares/cares/ares_cancel.o`
make: *** [/jenkins/workspace/gRPC_pull_requests_macos/workspace_c_macos_dbg_native/objs/dbg/third_party/cares/cares/ares__timeval.o] Error 1
make: *** Waiting for unfinished jobs....
cc -Ithird_party/protobuf/src -Ithird_party/googletest/googletest/include -Ithird_party/googletest/googlemock/include -Ithird_party/boringssl/include -Ithird_party/cares -Ithird_party/cares/cares -g -Wall -Wextra -Werror -Wno-long-long -Wno-unused-parameter -DOSATOMIC_USE_INLINED=1 -O0 -fPIC -I. -Iinclude -I/jenkins/workspace/gRPC_pull_requests_macos/workspace_c_macos_dbg_native/gens -I/opt/local/include -I/usr/local/include -D_DEBUG -DDEBUG -DINSTALL_PREFIX=\"/usr/local\" -DGRPC_TEST_SLOWDOWN_MACHINE_FACTOR=1.000000    -Ithird_party/cares -Ithird_party/cares/cares  -Ithird_party/cares/config_darwin -fvisibility=hidden -D_GNU_SOURCE -DWIN32_LEAN_AND_MEAN -D_HAS_EXCEPTIONS=0 -DNOMINMAX -DHAVE_CONFIG_H -std=c99 -Wsign-conversion -Wconversion -Wshadow -Wextra-semi   -Wno-sign-conversion -Wno-shorten-64-to-32 -Wno-invalid-source-encoding -MMD -MF /jenkins/workspace/gRPC_pull_requests_macos/workspace_c_macos_dbg_native/objs/dbg/third_party/cares/cares/ares_cancel.dep -c -o /jenkins/workspace/gRPC_pull_requests_macos/workspace_c_macos_dbg_native/objs/dbg/third_party/cares/cares/ares_cancel.o third_party/cares/cares/ares_cancel.c

2017-08-28 11:47:33,531 FAILED: make [ret=2, pid=11412]
2017-08-28 11:47:33,531 FAILED: Some tests failed
```

Also, any idea how to regenerate on freebsd?

**Comment by jtattermusch** [comment]
@mehrdada  it looks like config_freebsd/ares_config.h was added by you in https://github.com/grpc/grpc/pull/10724.  How do I regenerate it?

**Comment by mehrdada** [comment]
@jtattermusch If I remember correctly, I tried running ./configure for ares on a FreeBSD machine and extracted the generated `ares_config.h`.

**Comment by mehrdada** [comment]
@jtattermusch If you like, I can try generating a new one and push to your branch from a FreeBSD image?

**Comment by jtattermusch** [comment]
That would be great, thanks. Either push there or just send me the config
file.

**Comment by mehrdada** [comment]
https://github.com/jtattermusch/grpc/pull/10

**Comment by mehrdada** [comment]
@y-zeng sorry I missed your previous comment above. I already regenerated it and @jtattermusch has merged. I have a FreeBSD machine at home, but just in case for future reference, you can create a GCE machine with a FreeBSD image with:

```
gcloud compute instances create "freebsd11-machine-name" \
--machine-type "n1-standard-8" \
--maintenance-policy "MIGRATE" \
--image "freebsd-11-0-release-p1-amd64" --image-project=freebsd-org-cloud-dev \
--boot-disk-size "40" --boot-disk-type "pd-ssd"
```

**Comment by y-zeng** [comment]
Thanks @mehrdada!
@jtattermusch Please run the Jenkins tests with `GRPC_DNS_RESOLVER=ares` to make sure the upgrade will not break the resolver.

**Comment by jtattermusch** [comment]
Python 32 bit windows artifact compiled with mingw still broken:
```
In file included from third_party/cares/cares/ares_init.c:58:0:
third_party/cares/cares/ares_library_init.h:31:30: error: expected declaration specifiers or '...' before '*' token
 typedef NETIO_STATUS (WINAPI *fpGetBestRoute2_t) ( NET_LUID *, NET_IFINDEX, const SOCKADDR_INET *, const SOCKADDR_INET *, ULONG, PMIB_IPFORWARD_ROW2, SOCKADDR_INET * );
                              ^
In file included from third_party/cares/cares/ares_init.c:58:0:
third_party/cares/cares/ares_library_init.h:38:8: error: unknown type name 'fpGetBestRoute2_t'
 extern fpGetBestRoute2_t ares_fpGetBestRoute2;
        ^~~~~~~~~~~~~~~~~
third_party/cares/cares/ares_init.c: In function 'getBestRouteMetric':
third_party/cares/cares/ares_init.c:1022:3: error: unknown type name 'MIB_IPFORWARD_ROW2'
   MIB_IPFORWARD_ROW2 row;
   ^~~~~~~~~~~~~~~~~~
third_party/cares/cares/ares_init.c:1025:6: error: called object 'ares_fpGetBestRoute2' is not a function or function pointer
      ares_fpGetBestRoute2(/* The interface to use.  The index is ignored since we are
      ^~~~~~~~~~~~~~~~~~~~
In file included from third_party/cares/cares/ares_init.c:58:0:
third_party/cares/cares/ares_library_init.h:38:26: note: declared here
 extern fpGetBestRoute2_t ares_fpGetBestRoute2;
                          ^~~~~~~~~~~~~~~~~~~~
third_party/cares/cares/ares_init.c:1042:12: error: request for member 'Metric' in something not a structure or union
      || row.Metric == (ULONG)-1
            ^
third_party/cares/cares/ares_init.c:1043:12: error: request for member 'Metric' in something not a structure or union
      || row.Metric > ((ULONG)-1) - interfaceMetric) {
            ^
third_party/cares/cares/ares_init.c:1054:13: error: request for member 'Metric' in something not a structure or union
   return row.Metric + interfaceMetric;
             ^
third_party/cares/cares/ares_init.c: In function 'get_SuffixList_Windows':
third_party/cares/cares/ares_init.c:1430:32: warning: passing argument 1 of 'next_suffix' from incompatible pointer type [-Wincompatible-pointer-types]
       while (len = next_suffix(&pp, len))
                                ^
third_party/cares/cares/ares_init.c:1347:15: note: expected 'const char **' but argument is of type 'char **'
 static size_t next_suffix(const char** list, const size_t advance)
               ^~~~~~~~~~~
third_party/cares/cares/ares_init.c:1430:7: warning: suggest parentheses around assignment used as truth value [-Wparentheses]
       while (len = next_suffix(&pp, len))
       ^~~~~
third_party/cares/cares/ares_init.c: In function 'getBestRouteMetric':
third_party/cares/cares/ares_init.c:1055:1: warning: control reaches end of non-void function [-Wreturn-type]
 }
 ^
C:\tools\msys64\mingw32\bin\gcc.exe -mdll -O -Wall -IC:\Python27_32bits\include -IC:\Python27_32bits\PC -c t:\tmp\tmpcumiuy\a.c -o t:\tmp\tmpcumiuy\a.o
```

**Comment by jtattermusch** [comment]
@y-zeng what exactly do you mean by "GRPC_DNS_RESOLVER=ares to make sure the upgrade will not break the resolver."? We are currently not running those tests by default? That seems like a problem.

I can run tests with GRPC_DNS_RESOLVER=ares locally, but it sounds like a bad idea to only do it as a one-off instead of running them continuously.

**Comment by jtattermusch** [comment]
@y-zeng   ok, I see, the GRPC_DNS_RESOLVER=ares  tests are being run as part of linux portability suite (and only runs as build_only on PR) https://github.com/grpc/grpc/blob/f769bf12737bcff22b98b7265c5d4cd6c7bebd4d/tools/run_tests/run_tests_matrix.py#L245.

**Comment by jtattermusch** [comment]
@y-zeng   I ran the GRPC_DNS_RESOLVER=ares tests for C and C++ locally (with --use_docker)
and I only got this one failure which seems unrelated:
```
2017-08-30 15:58:42,392 FAILED: bins/dbg/json_run_localhost --scenarios_json '{"scenarios": [{"name": "cpp_protobuf_async_client_unary_1channel_64wide_128Breq_8MBresp_insecure", "warmup_seconds": 0, "benchmark_seconds": 1, "num_servers": 1, "server_config": {"async_server_threads": 0, "channel_args": [{"str_value": "latency", "name": "grpc.optimization_target"}, {"int_value": 1, "name": "grpc.minimal_stack"}], "security_params": null, "threads_per_cq": 0, "server_type": "ASYNC_SERVER"}, "num_clients": 1, "client_config": {"security_params": null, "channel_args": [{"str_value": "latency", "name": "grpc.optimization_target"}, {"int_value": 1, "name": "grpc.minimal_stack"}], "async_client_threads": 1, "outstanding_rpcs_per_channel": 1, "rpc_type": "UNARY", "payload_config": {"simple_params": {"resp_size": 8388608, "req_size": 128}}, "client_channels": 1, "threads_per_cq": 0, "load_params": {"closed_loop": {}}, "client_type": "ASYNC_CLIENT", "histogram_params": {"max_possible": 60000000000.0, "resolution": 0.01}}}]}' GRPC_POLL_STRATEGY=epoll1
```

I still need to fix the Python 32-bit artifact build on Windows, but after that, this PR should be good to go.

**Comment by apolcyn** [comment]
I ran this locally with not-checked-in-tests from PR #12210 and can confirm this doesn't break those

**Comment by jtattermusch** [comment]
Artifact build here:
https://grpc-testing.appspot.com/view/Artifacts/job/gRPC_build_artifacts/769/

**Comment by jtattermusch** [comment]
https://grpc-testing.appspot.com/view/Artifacts/job/gRPC_build_artifacts/776/

**Comment by jtattermusch** [comment]
I think the last remaining issue is the ruby gem artifact build failure:
https://sponge.corp.google.com/target?id=d36fe2ce-20f8-4ab8-91b5-30452cdec6b9&target=github/grpc&searchFor=&show=ALL&sortBy=STATUS

It's actually an warning coming form c-ares build, https://github.com/c-ares/c-ares/issues/143, but the problem is that our ruby-compiler-dock  docker image currently has  mingw version 4.7, which doesn't allow me silence the warning with -Wno-incompatible-pointer-types as this option was introduces in gcc 5+. Would updating the ruby-compiler-dock Dockerfile help?

**Comment by apolcyn** [comment]
the Docker file in the parent repo appears to have updated to mingw in https://github.com/rake-compiler/rake-compiler-dock/commit/abc54f570d797b4743579d79da1e7681ef925b26#diff-3254677a7917c6c01f55212f86c57fbf. looking into this

**Comment by jtattermusch** [comment]
@apolcyn  yes, I think updating ruby-compiler-dock would help but there's no clear instructions on how to arrive from the vanilla version for the Dockerfile to our patched version.  @nicolasnoble ?

**Comment by apolcyn** [comment]
PTAL https://github.com/jtattermusch/grpc/pull/12, which is a work in progress on a manual update

**Comment by apolcyn** [comment]
FWIW, https://github.com/jtattermusch/grpc/pull/12 seems to build the gems succesfully now

**Comment by apolcyn** [comment]
Looking further into it, I believe https://github.com/jtattermusch/grpc/pull/12 can work.

After building the rake-compiler-dock Dockerfile based off of ubuntu:17.04 (as is done in the parent repo), pulling down the debian gcc/mingw packages gets version 6.2, and allows disabling those compiler warnings.

However, ubuntu:17.04 also comes with openssl version 1.0.2 system libs installed, and so I needed to add the new `EMBED_OPENSSL=true` (the old docker file only had openssl 1.0.1 and so would not get picked up automatically by the makefile).

@nicolasnoble can you take a look at 
https://github.com/jtattermusch/grpc/pull/12?

**Comment by jtattermusch** [comment]
Known failures:
- https://github.com/grpc/grpc/issues/12313
- https://github.com/grpc/grpc/issues/9150
- https://github.com/grpc/grpc/issues/12461
- https://github.com/grpc/grpc/issues/12261
- https://github.com/grpc/grpc/issues/12312

**Pull request merged**